Added basic Expression classes

maxqchen opened the pull request:
A lot of details, declarations and implementations are missing because
of unresolved discussions (statements vs. assignment expressions,
primitive operations vs. macros, etc.) as well as VariableIdentifier not
holding a ref to Value (I'm guessing this is tba? or am I using it wrong?)

Throwing this up so we can get some discussion here regarding the above issues + feedback.

maxqchen committed:
Added basic Expression classes

A lot of details, declarations and implementations are missing because
of unresolved discussions (statements vs. assignment expressions,
primitive operations vs. macros, etc.) as well as VariableIdentifier not
holding a ref to Value for function calls.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/Expression.java:
```diff
@@ -0,0 +1,21 @@
+ *
+ */
+public abstract class Expression
+{
```
Would you mind putting the `{` on the same line as the declaration (here and everywhere else in the patch)? `public abstract class Expression {`

maxqchen replied:
Is there an eclipse formatter .xml I can import?

mtrberzi replied:
Ctrl-Shift-L seems to do most of what we want.

lucaswoj replied:
@maxqchen You'll have to address this before I can merge this pull request.

maxqchen replied:
Huh, I could've sworn my ctrl-shift-f went off. Will do.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/Expression.java:
```diff
@@ -0,0 +1,21 @@
+	// and then acting upon it. Would it be possible to remove one or more of these
+	// and replace them with exceptions?
+	public abstract boolean isCompiletimeEvaluable();
+	public abstract boolean isRuntimeEvaluable();
```
But during tree collapsing we will have to make that check. Basically, we will have a visitor that asks each node if it is `isCompiletimeEvaluable` and, if so, will replace it with a `LiteralExpression` of its `evaluate`-ed value.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/FunctionInvocationExpression.java:
```diff
@@ -0,0 +1,63 @@
+public class FunctionInvocationExpression extends Expression
+{
+	private Expression input;
+	private VariableIdentifier function;
```
I think `function` will also need to be an `Expression`. Since functions are first class objects, one could return a function from another function and immediate execute it, `foo()()` (as ugly as that is)

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/FunctionInvocationExpression.java:
```diff
@@ -0,0 +1,63 @@
+	@Override
+    public TypeValue resultType()
+    {
+		// TODO: VariableIdentifier doesn't specify a Value/FunctionValue?
```
No, you'd have to ask the scope for that.

mtrberzi replied:
Yeah, this requires a `Function` class to be derived from the variable, which can then be evaluated. You have to go through the scope to do it, but it's not quite something we support at the moment.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/Expression.java:
```diff
@@ -0,0 +1,21 @@
+	public abstract TypeValue resultType();
+	public abstract Value evaluate();
+	public abstract void verify();
+	public abstract boolean isAssignable();
```
Every Expression should also have a `Scope getScope()` method which could be implemented in this base class. This is the bridge between `VariableIdentifier`s and `Value`s.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/FunctionInvocationExpression.java:
```diff
@@ -0,0 +1,63 @@
+	{
+		// TODO: check that function is actually a function
+		// TODO: check that the function takes in the type of input (or do it in verify)?
+		// pending because VariableIdentifier doesn't seem to specify Value/FunctionValue
```
So, you should also take in a `Scope` as an argument here and then we can do all these checks in `verify`

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/FunctionInvocationExpression.java:
```diff
@@ -0,0 +1,63 @@
+    public Value evaluate()
+    {
+	    // TODO VariableIdentifier doesn't specify a Value/FunctionValue?
+	    return null;
```
All you need to do is defer the work to the (yet to be implemented) `FunctionValue` object 

``` java
return functionExpression.getValue().evaluate(inputExpression.getValue())
```

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/FunctionInvocationExpression.java:
```diff
@@ -0,0 +1,63 @@
+	@Override
+    public void verify()
+    {
+	    // TODO: checks in constr or here?
```
The checks will have to go here because of the way things get resolved within the `Scope`.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/FunctionInvocationExpression.java:
```diff
@@ -0,0 +1,63 @@
+	@Override
+    public boolean isAssignable()
+    {
+	    // TODO: how to determine this?
```
This will have to be handled case by case in the implementing classes. Function invocation isn't typically a valid lvalue so just return false here.

mtrberzi replied:
Recursively, but we can't do it yet.

mtrberzi commented on src/main/java/org/whdl/frontend/syntaxtree/LiteralExpression.java:
```diff
@@ -0,0 +1,55 @@
+	@Override
+	public void verify()
+	{
+		// Note (max) - anything else needed for this?
```
No, that's pretty much it. `LiteralExpression`s are simple, aren't they?

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/FunctionInvocationExpression.java:
```diff
@@ -0,0 +1,63 @@
+	@Override
+    public boolean isCompiletimeEvaluable()
+    {
+	    // TODO Auto-generated method stub
```
In both `isCompiletimeEvaluable` and `isRuntimeEvaluable` you'll probably have to do something like

``` java
return functionExpression.isCompiletimeEvaluable() &&  inputExpression.isCompiletimeEvaluable()
```

maxqchen replied:
Is it possible for an expression to be compile time evaluated to a function that isn't compile time evaluable?

Then we would need to also check against the resulted FunctionValue.

lucaswoj replied:
Yes, that would be possible.

I think we should hold off on fleshing out the `FunctionInvoactionExpression` until more dependencies are ready.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/FunctionInvocationExpression.java:
```diff
@@ -0,0 +1,63 @@
+ * @author Max
+ *
+ */
+public class FunctionInvocationExpression extends Expression
```
I'm not so sure, in retrospect, that this was a great class to start with. Lots of dependencies...

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/LiteralExpression.java:
```diff
@@ -0,0 +1,55 @@
+	@Override
+	public boolean isCompiletimeEvaluable()
+	{
+		return true;
```
I think you'll actually have to defer to the value's `isCompiletimeEvaluable` method -- for example, types are literals but not "runtime evaluable".

maxqchen replied:
This (in addition to my comment above for the function) seems to imply that we want `isCompileTimeEvaluable` and `isRuntimeEvaluable` as abstract methods of `Value` as well. Should we superclass all of these?

lucaswoj replied:
Correct. There is an outstanding pull request for that functionality in #36. 

Conceptually, I don't think `Expression` and `Value` should have a superclass but I might be convinced otherwise.

maxqchen replied:
Having them implement a common interface with those 2 methods would make more sense, actually.

lucaswoj replied:
I mean, yes, it makes sense in that they have methods with the same name (although I'm still not sure calling values `evaluable` is sensical) but I can't think of an instance in which you'd want to process them interchangeably.

TysonAndre replied:
@maxqchen side note: #36 changed the name of isRuntimeEvaluable to isSynthesizable.
When it is merged, this code would have to be changed as well.

mtrberzi commented on src/main/java/org/whdl/frontend/syntaxtree/FunctionInvocationExpression.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+	@Override
+    public boolean isRuntimeEvaluable()
```
this also becomes `isSynthesizable()`

mtrberzi commented:
This is mostly fine. Make corrections (mostly `isSynthesizable()` refactoring) and resolve conflicts, and I think I'm ready to merge.

maxqchen committed:
FunctionExpression Branch

maxqchen committed:
Expression && LiteralExpression
Modified tests to remove FacadeExpression

maxqchen committed:
Expression and LiteralExpression

- Moved FunctionInvocationExpression and pals to another branch, no
longer in this commit
- Changed some of the tests to use LiteralExpression
- Changed Value->TypeValue in TypeMismatchException

maxqchen commented:
I removed FunctionInvocationExpr from this since it needed a bit more work. This is just Expression/LiteralExpression now.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/Expression.java:
```diff
@@ -1,6 +1,17 @@
+ * Base expr class
+ * @author Max
+ *
+ */
```
I'm all for using docblocks but we should try to standardize our use across the source. You can commit whatever you have here and I'm going to create a new issue to retrofit the rest of the codebase. Lets discuss docblocks on task #50!

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/TypeMismatchException.java:
```diff
@@ -4,9 +4,9 @@
 
   private TypeValue expectedType;
-  private Value actualType;
+  private TypeValue actualType;
```
Not sure if you did this or it was a bad merge. I replaced `TypeValue` with `Value` here to handle the case where a non-type value is used as a type. Perhaps we should have a different exception in that case. I'll open task #51 for this. For this pull request, I'd prefer if you just reverted this file.

EDIT: fixed typo

maxqchen replied:
Yeah I'll revert it. Was looking through the code and figured this was a typo.

lucaswoj commented:
Almost ready to go :smile:

maxqchen committed:
Formatting
Removed doc blocks
Reverted TypeMismatchException

mtrberzi commented:
@lucaswoj does this look good to you? I'll sign off but I don't know if you needed to see anything else
